src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2299 @@
+        if prev_key_values is not None:
+            k = prev_cross_key
+            v = prev_cross_value
+
+        # if we're returning the cache for later use, store these now for later return (can be done without having prev_key_values provided)
+        if use_cache:
+            updated_cross_key = k
+            updated_cross_value = v
+
+        ctx_len = k.size(1)
+        # This is part of a workaround to get around fork/join parallelism
+        # not supporting Optional types.
+        if key_padding_mask is not None and key_padding_mask.dim() == 0:
+            key_padding_mask = None
+
+        if key_padding_mask is not None:
+            assert key_padding_mask.size(0) == bsz
+            assert key_padding_mask.size(1) == ctx_len
+
+        if self.attention_activation == "softmax":
+            attn_weights = self.softmax_attention(q, k, key_padding_mask, num_incremental_steps)
+        else:
+            attn_weights = self.element_attention(q, k, key_padding_mask, num_incremental_steps)
+
+        v = self.hidden_dropout(v, batch_first=True)
+        kernel = self.attention_dropout(attn_weights)
+        # B x L2 x D -> L2 x B x D
+        h = torch.bmm(kernel, v).transpose(0, 1)
+        # L2 x B x D
+        h = self.activation(self.h_proj(h * r))
+        h = self.dropout(h)
+        out = torch.addcmul(query, u, h - query)
+
+        if not self.prenorm:
+            out = self.norm(out)
+
+        outputs = (out, attn_weights) if output_attentions else (out,)
+        if use_cache:
+            outputs = outputs + (updated_cross_key, updated_cross_value)
+
+        return outputs
+
+
+# Positional embeddings
+# copied without modification from original Mega code
+class SimpleRelativePositionalBias(nn.Module):
+    def __init__(self, config: MegaConfig):
+        super().__init__()
+        self.config = config
+        self.max_positions = self.config.max_positions if self.config.chunk_size < 0 else self.config.chunk_size
+        self.rel_pos_bias = nn.Parameter(torch.Tensor(2 * config.max_positions - 1))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        std = self.config.initializer_range
+        nn.init.normal_(self.rel_pos_bias, mean=0.0, std=std)
+
+    def forward(self, seq_len):
+        if seq_len > self.max_positions:
+            raise ValueError("Sequence length {} going beyond max length {}".format(seq_len, self.max_positions))
+
+        # seq_len * 2 -1
+        b = self.rel_pos_bias[(self.max_positions - seq_len) : (self.max_positions + seq_len - 1)]
+        # seq_len * 3 - 1
+        t = F.pad(b, (0, seq_len))
+        # (seq_len * 3 - 1) * seq_len
+        t = torch.tile(t, (seq_len,))
+        t = t[:-seq_len]
+        # seq_len x (3 * seq_len - 2)
+        t = t.view(seq_len, 3 * seq_len - 2)
+        r = (2 * seq_len - 1) // 2
+        start = r
+        end = t.size(1) - r
+        t = t[:, start:end]
+        return t
+
+
+class RotaryRelativePositionalBias(nn.Module):
```

Comment:
Quick question, what is this based on? (we have already 3 implementations of Rotary position embedding, would be nice to have a reference (I mean is this from their repo etc)

Reply:
One example [here](https://github.com/ArthurZucker/transformers/blob/fix-zac-pipleine/src/transformers/models/esm/modeling_esm.py#L87)

Reply:
Apologies, that could have been clearer. This is based on the original implementation! See https://github.com/facebookresearch/mega/blob/09f1a872c77bdf78e1b128c21c002f08fd6cc2dc/fairseq/modules/relative_positional_bias.py#L49

Reply:
Edit, after reading the paper, it seems that the `relative positional
bias` is based on the `T5` , if we can have copied from would be great! Otherwise let's just add MegaRotaryRelativePositionalBias

Reply:
Also let's move them to the top of the file 😉